Flesh out documentation and switch to using patches

gene1wood opened the pull request:
This adds all details to go from a brand new CentOS 7 instance to a built RPM
Updates to use a patch model instead of a fork of Auth0's code

Also Update Makefile to allow for tgz or zip file
Changes to use Auth0's code and then apply a patch.
Also creates a `make setup` target which sets up the system to build the RPM

Once this PR is merged we should delete the https://github.com/mozilla-iam/ad-ldap-connector repo as it's no longer used

gene1wood committed:
Update Makefile to allow for tgz or zip file

gene1wood committed:
Flesh out documentation to fully describe how to use the tool

This adds all details to go from a brand new CentOS 7 instance to a built RPM

april commented on README.md:
```diff
@@ -1,34 +1,80 @@
+
+# Ideally
+
+The upstream package would need modifications to properly install without 
```
I don't understand this? It would, or would not?

gene1wood replied:
Ya, I also didn't understand what Guillaume meant here, I was thinking of removing it but couldn't think of what to replace it with, maybe just delete it.

gene1wood committed:
Update to use a patch model instead of a fork of Auth0's code

Changes to use Auth0's code and then apply a patch.
Also creates a `make setup` target which sets up the system to build the RPM

april commented on Makefile:
```diff
@@ -43,32 +43,45 @@ fpm: extract npm_verify
+	sudo yum install -y git unzip rpm-build nodejs gcc gcc-c++ patch autoconf automake bison libffi-devel libtool readline-devel sqlite-devel zlib-devel openssl-devel
+	gpg2 --keyserver hkp://pool.sks-keyservers.net --recv-keys 409B6B1796C275462A1703113804BB82D39DC0E3 7D2BAF1CF37B13E2069D6956105BD0E739499BDB
+	test -e ~/.rvm/scripts/rvm || curl -sSL https://get.rvm.io | bash -s stable
+	~/.rvm/bin/rvm install 2.7.2
```
Should this version number be a variable at the top?

gene1wood replied:
Good idea, I'll do that.

gene1wood replied:
I've implemented your suggestion in 42fa0e560d437768800c711c10412bb37634c98a

april commented on README.md:
````diff
@@ -1,34 +1,73 @@
+       1:ad-ldap-connector-1.2.3_mozilla-################################# [100%]
+    ```
+- Change the file `/opt/ad-ldap-connector/environ` if you use a proxy
+- Copy over your previous certs directory and `config.json`. If you have no 
````
Does this not need to happen prior to running systemctl?

gene1wood replied:
Ya, the text above talking about systemctl is output from the actual RPM installation process, which is saying the same thing (configure `config.json` before starting the process). Any suggestions on what to change this to?

april replied:
Ohhh, I see, it's just saying that's the output, not to run them. Sorry, I missed that context when looking at the commit.

april commented on patches/001_email-verified.patch:
```diff
@@ -0,0 +1,39 @@
++  if (raw_data['zimbraAlias']) {
++    all_emails = all_emails.concat(raw_data['zimbraAlias']);
++  }
++
```
I would probably make this entire thing:

```javascript
const all_emails = [].concat(raw_data.mail || [],
                             raw_data.zimbraAlias || [])
```

gene1wood replied:
Right, I figured I'd just capture the existing code that was in the forked repo in this patch file for the moment and worry about improving the code after we'd cutover to this patch based deployment method. 

I captured this suggestion in #7

april commented on patches/001_email-verified.patch:
```diff
@@ -0,0 +1,39 @@
+     displayName: raw_data.displayName,
+@@ -8,7 +17,7 @@ module.exports = function (raw_data) {
+     },
+     nickname: raw_data['sAMAccountName'] || raw_data['cn'] || raw_data['commonName'],
```
This code is mixing dot and bracket notation. Use only dot notation (`raw_data.sAMAccountName`) unless looking up by variable.

gene1wood replied:
This is Auth0's code which I'd like to keep intact and change as little as possible. To improve this we'd want to PR to their upstream : https://github.com/auth0/ad-ldap-connector/

gene1wood committed:
Abstract the ruby version into a Makefile variable
